backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt

```diff
@@ -1272,6 +1275,94 @@ class SubmissionDatabaseService(
+                                    SequenceEntriesTable.versionColumn
+                                )
+                    }
+                    .select(
+                        SequenceEntriesPreprocessedDataTable.accessionColumn,
+                        SequenceEntriesPreprocessedDataTable.versionColumn,
+                        SequenceEntriesPreprocessedDataTable.pipelineVersionColumn,
+                    )
+                    .where {
+                        (SequenceEntriesPreprocessedDataTable.pipelineVersionColumn less earliestVersionToKeep) and
+                            (SequenceEntriesTable.organismColumn eq organism)
+                    }
+                    .limit(batchSize)
+                    .map {
+                        Triple(
+                            it[SequenceEntriesPreprocessedDataTable.accessionColumn],
+                            it[SequenceEntriesPreprocessedDataTable.versionColumn],
+                            it[SequenceEntriesPreprocessedDataTable.pipelineVersionColumn],
+                        )
+                    }
+                SequenceEntriesPreprocessedDataTable.deleteWhere {
+                    Triple(
+                        SequenceEntriesPreprocessedDataTable.accessionColumn,
+                        SequenceEntriesPreprocessedDataTable.versionColumn,
+                        SequenceEntriesPreprocessedDataTable.pipelineVersionColumn,
+                    ) inList keysToDelete
+                }
+            }
+            totalDeletedRows += deletedRowsInLastBatch
+            log.debug {
+                "Deleted $deletedRowsInLastBatch rows in this batch for organism '$organism' " +
+                    "with version < $earliestVersionToKeep"
+            }
+        } while (deletedRowsInLastBatch == batchSize)
+        if (totalDeletedRows > 0) {
+            log.info(
+                "Finished cleanup of outdated preprocessing data for organism '$organism', " +
+                    "with version < $earliestVersionToKeep. Total deleted rows: $totalDeletedRows.",
+            )
+        } else {
+            log.info(
+                "No outdated preprocessing data found for organism '$organism' with version < $earliestVersionToKeep.",
+            )
+        }
+    }
+
+    /**
+     * Garbage collects outdated preprocessed entries for all organisms.
+     * This is a convenience method that iterates over all distinct organisms in the database
+     * and calls [cleanUpOutdatedPreprocessingData] for each organism.
+     */
+    fun cleanUpOutdatedPreprocessingDataForAllOrganisms() {
+        CurrentProcessingPipelineTable.selectAll()
+            .forEach { row ->
+                cleanUpOutdatedPreprocessingData(
+                    organism = row[CurrentProcessingPipelineTable.organismColumn],
+                    earliestVersionToKeep = row[CurrentProcessingPipelineTable.versionColumn],
```

Comment:
In https://github.com/loculus-project/loculus/pull/4584 we only deleted versions that were less than the currentVersion -1, here we now delete any version less than the currentVersion - could this ever be an issue?

If not lets rename the variable `earliestVersionToKeep` to `currentPipelineVersion` to be more explicit.

Reply:
we don't keep `-1` around anymore in this version, see PR description for reasoning:
> Clean up all outdated prepro versions (deciding not to keep the latest unused one around as it creates a lot of db load to have twice the number of versions around)